### PR TITLE
Create a Harvester Manager plugin

### LINKED ISSUE #6790: Move Harvester Management to its own package
- After https://github.com/rancher/dashboard/issues/5630 is resolved we should move out the harvester manager related code from shell to it's own package
- This package will always be bundled with the dashboard

### PR DESCRIPTION
**Pull request by mantis-toboggan-md**
### Summary
Fixes #6790, #7256

This PR moves code related to importing or 'managing' harvester clusters into an extension, to be bundled with the rancher dashboard. There should be no changes to functionality here, just code moving. 

### Areas or cases that should be tested
Importing harvester clusters, navigating into a harvester cluster ie loading the Harvester pkg/extension:
- harvester cluster without bundled extension (pre 1.1.0 harvester) offline-preferred and remote-preferred
- harvester cluster that does support extension (1.1.0) offline-preferred and remote-preferred

### CODE REVIEW

File: pkg/harvester-manager/pages/c/_cluster/index.vue [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <script>
-import { NAME } from '@shell/config/product/harvester-manager';
+import { NAME } from '~/pkg/harvester-manager/config/harvester-manager';
```
Comment (richard-cox) [inline pkg/harvester-manager/pages/c/_cluster/index.vue:2]: `~` will break the currently disabled the `check-plugin` gate

File: shell/edit/provisioning.cattle.io.cluster/import.vue [hunk 2]
```diff
@@ -10,7 +10,7 @@ import { CAPI, HCI } from '@shell/config/types';
 import ClusterMembershipEditor from '@shell/components/form/Members/ClusterMembershipEditor';
 import { Banner } from '@components/Banner';
 import { canViewClusterMembershipEditor } from '@shell/components/form/Members/ClusterMembershipEditor.vue';
-import { NAME as HARVESTER_MANAGER } from '@shell/config/product/harvester-manager';
+import { NAME as HARVESTER_MANAGER } from '~/pkg/harvester-manager/config/harvester-manager';
```
Comment (richard-cox) [inline shell/edit/provisioning.cattle.io.cluster/import.vue:13]: `~`

File: shell/config/features.js [hunk 3]
```diff
@@ -1,3 +1,5 @@
 export const ONE_WAY = [
   'token-hashing', 'multi-cluster-management'
 ];
+
+export const HARVESTER_NAME = 'harvester';
```
Comment (richard-cox) [inline shell/config/features.js:5]: It's scary how often this is still used outside of pkgs. At some point we'll need to address, not now though

**Review by richard-cox — CHANGES_REQUESTED** [review]
- At some point we need to tackle providing machine configs via plugins. I think @aalves08 might even have this working. If this is straight forward it would be good to do it here, if not then we should capture it in a PR
- There's some linting errors, probably worth a rebase to capture the new linting rules

Once these are all resolved i'll come back and give it a proper test

**Comment by aalves08** [comment]
> * At some point we need to tackle providing machine configs via plugins. I think @aalves08 might even have this working. If this is straight forward it would be good to do it here, if not then we should capture it in a PR
> * There's some linting errors, probably worth a rebase to capture the new linting rules
> 
> Once these are all resolved i'll come back and give it a proper test

That is done here: https://github.com/rancher/elemental-ui/blob/main/pkg/elemental/edit/provisioning.cattle.io.cluster/MachinePool.vue#L113-L115

but it's not the proper way of doing it because it's overriding the `MachinePool` file on `rancher/dashboard` once the extension installed, which may lead to problems like these which we are trying to avoid (and I still need to fix those on Elemental):
https://suse.slack.com/archives/C02DE1LRG65/p1668706876100059

File: shell/config/features.js (hunk 3 above)
Comment (mantis-toboggan-md) [reply shell/config/features.js:5]: Yeah, unfortunately I think we've already grabbed the low-hanging fruit and remaining references will take more work to remove.

**Comment by mantis-toboggan-md** [comment]
> > * At some point we need to tackle providing machine configs via plugins. I think @aalves08 might even have this working. If this is straight forward it would be good to do it here, if not then we should capture it in a PR
> > * There's some linting errors, probably worth a rebase to capture the new linting rules
> > 
> > Once these are all resolved i'll come back and give it a proper test
> 
> That is done here: https://github.com/rancher/elemental-ui/blob/main/pkg/elemental/edit/provisioning.cattle.io.cluster/MachinePool.vue#L113-L115
> 
> but it's not the proper way of doing it because it's overriding the `MachinePool` file on `rancher/dashboard` once the extension installed, which may lead to problems like these which we are trying to avoid (and I still need to fix those on Elemental): https://suse.slack.com/archives/C02DE1LRG65/p1668706876100059

@aalves08 is there a gh issue to track this? Plugins _should_ be able to use their own `machine-config` (and `cloud-credential`) folders but it looks like we'll need to update these getters to account for that (somehow): https://github.com/rancher/dashboard/blob/master/shell/store/plugins.js#L160-L167

File: pkg/harvester-manager/pages/c/_cluster/index.vue (hunk 1 above)
Comment (mantis-toboggan-md) [reply pkg/harvester-manager/pages/c/_cluster/index.vue]: done

File: shell/edit/provisioning.cattle.io.cluster/import.vue (hunk 2 above)
Comment (mantis-toboggan-md) [reply shell/edit/provisioning.cattle.io.cluster/import.vue]: done

**Comment by aalves08** [comment]
> 

@mantis-toboggan-md I don't have a gh issue to track this

**Review by richard-cox — CHANGES_REQUESTED** [review]
- Changes look good, unfortunately i couldn't test as all my DO harvester environments have gone and i can't seem to access anything on fremont. Once we get through the review our side would you be able to ask if someone harvester side could validate functionality the changes are ok?
- The updates to machine-config and credentials don't breaking anything non-harvester side, but noticed HarvesterServiceAddOnConfig. Couldn't see how this code wise links in, but i think it relates to rke2 cluster add on config... couldn't test this though. If it does relate, we'd probably should make it generic too as per machine-config

**Comment by richard-cox** [comment]
@mantis-toboggan-md This PR needs some TLC

**Comment by mantis-toboggan-md** [comment]
> * The updates to machine-config and credentials don't breaking anything non-harvester side, but noticed HarvesterServiceAddOnConfig. Couldn't see how this code wise links in, but i think it relates to rke2 cluster add on config... couldn't test this though. If it does relate, we'd probably should make it generic too as per machine-config

I'll look into this today and see if it can be implemented in a more generic way. `HarvesterServiceAddonConfig` is an...unusual component. It's used within the explorer's service edit page when the cluster being explored is provisioned on a harvester cluster.

**Comment by mantis-toboggan-md** [comment]
Per teams conversation @aalves08 will be working on a mechanism to allow plugins to provide their own tabs within ResourceTabs; this should allow us to pull the HarvesterAddonConfig out of /shell but there's no need to delay this PR in order to leverage that. I've made an issue to track this work here https://github.com/rancher/dashboard/issues/7774

**Comment by mantis-toboggan-md** [comment]
@WuJun2016 or @n313893254 could one (or both) of you take a look at this PR and verify that it doesn't break the Harvester cluster import functionality? Thank you

**Comment by WuJun2016** [comment]
@mantis-toboggan-md   LGTM,   I tested it with v1.1.0, latest, and v1.0.3, and everything is as expected.  thanks

**Review by richard-cox — DISMISSED** [review]

**Review by richard-cox — DISMISSED** [review]

**Review by richard-cox — APPROVED** [review]

**Comment by krumware** [comment]
harvester-manager imports are broken in the extensions starter: https://github.com/rancher/dashboard/issues/8008

**Pull request merged**